src/OpenTelemetry/Metrics/MetricProvider.cs

```diff
@@ -0,0 +1,147 @@
+
+            this.listener.Start();
+
+            this.cts = new CancellationTokenSource();
+
+            var token = this.cts.Token;
+            this.observerTask = Task.Run(async () =>
+            {
+                while (!token.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await Task.Delay(this.options.ObservationPeriodMilliseconds, token);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                    }
+
+                    this.listener.RecordObservableInstruments();
+                }
+            });
+        }
+
+        public Meter GetMeter(string name, string version)
+        {
+            var meter = new Meter(name, version);
+            this.meters.TryAdd(meter, 0);
+
+            return meter;
+        }
+
+        public void Dispose()
+        {
+            this.cts.Cancel();
+            this.observerTask.Wait();
+        }
+
+        internal void InstrumentPublished(Instrument instrument, MeterListener listener)
+        {
+            bool isInclude = false;
+
+            if (this.options.IncludeMeters != null && this.options.IncludeMeters.Length > 0)
+            {
+                foreach (var meterFunc in this.options.IncludeMeters)
+                {
+                    if (meterFunc(instrument))
+                    {
+                        isInclude = true;
+                        break;
+                    }
+                }
+            }
+            else
+            {
+                isInclude = this.meters.TryGetValue(instrument.Meter, out var _);
+            }
+
+            if (isInclude)
+            {
+                // Enable this Instrument if it should be included.
+                listener.EnableMeasurementEvents(instrument, null);
+
+                if (this.options.Verbose)
+                {
+                    Console.WriteLine($"Instrument {instrument.Meter.Name}:{instrument.Name} published.");
+                }
+            }
+        }
+
+        internal void MeasurementsCompleted(Instrument instrument, object? state)
+        {
+            if (this.options.Verbose)
+            {
+                Console.WriteLine($"Instrument {instrument.Meter.Name}:{instrument.Name} completed.");
+            }
+        }
+
+        internal void MeasurementRecorded<T>(Instrument instrument, T value, ReadOnlySpan<KeyValuePair<string, object?>> attribs, object? state)
+        {
+            if (this.options.Verbose)
+            {
+                Console.WriteLine($"Instrument {instrument.Meter.Name}:{instrument.Name} recorded {value}.");
+            }
+        }
+
+        internal class BuildOptions
+        {
+            public Func<Instrument, bool>[] IncludeMeters { get; set; } = new Func<Instrument, bool>[0];
+
+            public bool Verbose { get; set; } = true;
+
+            public int ObservationPeriodMilliseconds { get; set; } = 1000;
```

Comment:
Not a blocker, I think we need to call out that this is not observation period, instead, it is the delay/sleep between two observations.
Let's say each observation callback takes 800ms, and `ObservationPeriodMilliseconds` is 200ms, we will end up with pushes at every 1000ms (instead of every 200ms) approximately.

Reply:
Consider something similar to `scheduledDelayMillis` in the tracing spec https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#built-in-span-processors

Reply:
And this https://github.com/open-telemetry/opentelemetry-dotnet/blob/98055871511d0b56df75cb7443a1ed9a49918014/src/OpenTelemetry/BatchExportProcessor.cs#L58

Note that in OTel.NET it was called "Milliseconds" instead of "Millis" since the former is more idiomatic to .NET (and the later is more Java-ish).